### DISCUSSION TITLE
Writing impure reducers?

### DISCUSSION THREAD

**Discussion by mcrowder65** [pmndrs/zustand#697] [General]
Hello! 

We are adding zustand to our application to replace our use of React.useReducer.

One thing that bugs me about useReducer is it expects the reducers to **always** be pure even if super inefficient.

example:

```js
function reducer(state, action) {
   if(action.type === "SET_REQUIRED") {
    return {
      ...state,
      required: true
    }
   }
  return state
}
```

That spread of state could be a huge operation if state is huge.

What I'm finding from initial tests is that I can just do 

```js
function reducer(state, action) {
   if(action.type === "SET_REQUIRED") {
     state.required = true
     return state
   }
  return state
}
```

I'm not running into any issues with this, am I missing something?

**Comment by dai-shi** [comment]
That may seem to work in zustand, because it merges states.
But, mutating old state may cause issues in complex scenarios, so not recommended.

> That spread of state could be a huge operation if state is huge.

Are you sure if it's a huge operation? Do you have thousands of properties? meaning `Object.keys(state).length`.
I would say that should be revised and make it more organized. It's not only the limitation of zustand, but JS in general.

**Comment by mcrowder65** [comment]
Hey @dai-shi, thanks so much for the reply.

That was a pretty contrived example.

Our application is a document editor where we have an array of sections, and each section has an array of blocks.

So when we update something as simple as the block name, we have to map over all the sections, find the matching section, then map over the section's blocks and find the matching block, and update that block.

```js
export const updateBlock = (state, action, updateFn) => {
  return {
    ...state,
    hasChanges: true,
    data: {
      ...state.data,
      sections: state.data.sections.map((section, sectionIndex) => {
        if (sectionIndex === state.sectionIndex) {
          return {
            ...section,
            blocks: section.blocks.map((block) => {
              if (
                block.id === action.blockId ||
                block.id === action.block?.id
              ) {
                return updateFn(block)
              }
              return block
            }),
          }
        }
        return section
      }),
    },
  }
}
```

Could we alternatively write it as follows:
```js
export const updateBlock = (state, action, updateFn) => {
   let block = state.data.sections[state.sectionIndex].blocks[action.blockIndex]
   state.data.sections[state.sectionIndex].blocks[action.blockIndex] =updateFn(block)
   state.hasChanges = true
   return state
}
```

Thanks again

**Reply by dai-shi** [reply]
Okay, I misunderstood what "a huge operation" means. I mistook it's about performance concern.

Please check this wiki page: https://github.com/pmndrs/zustand/wiki/Updating-nested-state-object-values

You might be interested in [valtio](https://github.com/pmndrs/valtio) if you like the mutation style.
